### DISCUSSION TITLE
whether can save uefi environment when memtest run

### DISCUSSION THREAD

**Discussion by changpenghui616** [memtest86plus/memtest86plus#247] [Q&A]
In current memtest86+ framework, after call efi_get_memory_map, call exit_boot_services to release uefi resource ，
so, whether can not call exit_boot_services to save uefi environment. of course， uefi memory be reserve and not test.
thanks

**Comment by x86fr** [comment]
No. UEFI can allocate a LOT of memory, and as a memory testing tool, we want to test **all the memory available**. We don't want to skip the memory reserved by any operating system or by the UEFI Framework. That's why we coded our own drivers for USB for example.

**Reply by changpenghui616** [reply]
I understand the memtest86+ target is test all memory， i want to confirm this framework whether  support this operate（save uefi boot service memory），reduce the test space to do something else. 
I try to  set all memory type as 'E820_RESERVED' ,  the result is normal. have some message.
but when I delete ‘exit_boot_services’ in efisetup.c,  it cant work(black screen), so i want know what happend, thanks

**Comment by yoshinoyatoko** [comment]
Maybe you can call UEFI Runtime Variable service to save some environment info into nvram(such as: SPIROM chip)

**Reply by changpenghui616** [reply]
i want to use bootservice proctol in test runing，it look cant work now, so i need confirm why cant work if havnt exit_boot_service
